src/util/string.cpp

```diff
@@ -13,4 +13,42 @@ void ReplaceAll(std::string& in_out, const std::string& search, const std::strin
     if (search.empty()) return;
     in_out = std::regex_replace(in_out, std::regex(search), substitute);
 }
+
+LineReader::LineReader(std::span<const std::byte> buffer, size_t max_read)
+    : start(buffer.begin()), end(buffer.end()), max_read(max_read), it(buffer.begin()) {}
+
+std::optional<std::string> LineReader::ReadLine()
+{
+    if (it == end) {
+        return std::nullopt;
+    }
+
+    auto line_start = it;
+    std::string line{};
+    while (it != end) {
+        char c = static_cast<char>(*it);
+        line += c;
+        ++it;
+        if (c == '\n') break;
+        if ((size_t)std::distance(line_start, it) >= max_read) throw std::runtime_error("max_read exceeded by LineReader");
+    }
+
+    line = TrimString(line); // delete trailing \r and/or \n
+    return line;
+}
+
+// Ignores max_read but won't overflow
+std::string LineReader::ReadLength(size_t len)
+{
+    if (len == 0) return "";
+    if (Left() < len) throw std::runtime_error("Not enough data in buffer");
+    std::string out(reinterpret_cast<const char*>(&(*it)), len);
```

Comment:
Can we return a `string_view` to the internal buffer?
(preventing allocation & copy)

Reply:
It seems to still return a `std::string`:
```
src/util/string.cpp:  std::string LineReader::ReadLength(size_t len)
src/util/string.h:    std::string ReadLength(size_t len);

```

Reply:
Oh I see I misunderstood. For now I want LineReader to return new string objects because it reads directly from the HTTPClient's `m_recv_buffer` which is then erased. I know we can prevent that copy by being more clever with `m_recv_buffer` and I still plan on exploring that, maybe as a follow up? It's nice and simple now and there is a clean separation between the socket layer receiving data and the `HTTPRequest`

Reply:
Sounds good, thanks for the context :)